lake-primitives/src/types/receipts.rs

```diff
@@ -148,113 +151,687 @@ impl Action {
     pub fn signer_public_key(&self) -> PublicKey {
         self.signer_public_key.clone()
     }
+}
+
+pub trait Action {
+    fn metadata(&self) -> &ActionMetadata;
+
+    fn receipt_id(&self) -> CryptoHash {
+        self.metadata().receipt_id()
+    }
+    fn predecessor_id(&self) -> AccountId {
+        self.metadata().predecessor_id()
+    }
+    fn receiver_id(&self) -> AccountId {
+        self.metadata().receiver_id()
+    }
+    fn signer_id(&self) -> AccountId {
+        self.metadata().signer_id()
+    }
+    fn signer_public_key(&self) -> PublicKey {
+        self.metadata().signer_public_key()
+    }
+
+    fn as_create_account(&self) -> Option<&CreateAccount> {
+        None
+    }
+
+    fn as_deploy_contract(&self) -> Option<&DeployContract> {
+        None
+    }
+
+    fn as_function_call(&self) -> Option<&FunctionCall> {
+        None
+    }
+
+    fn as_transfer(&self) -> Option<&Transfer> {
+        None
+    }
+
+    fn as_stake(&self) -> Option<&Stake> {
+        None
+    }
+
+    fn as_add_key(&self) -> Option<&AddKey> {
+        None
+    }
+
+    fn as_delete_key(&self) -> Option<&DeleteKey> {
+        None
+    }
+
+    fn as_delete_account(&self) -> Option<&DeleteAccount> {
+        None
+    }
+
+    fn as_delegate_action(&self) -> Option<&DelegateAction> {
+        None
+    }
```

Comment:
These do not belong here, they should be only implemented on the Action type (btw, I believe `ActionKind` enum should be named just `Action`, and this `Action` trait should be `ActionMetadata`)